Anti tip

brownkat6 opened the pull request:
Added methods to Drive so that when the Robot tips past a certain threshold (currently 5 degrees), the drivetrain modifies the motor speeds to compensate for the tipping. It does not override driver control. Also added a button to disable/enable the Anti-Tip functions.

brownkat6 committed:
Added AntiTip Command framework

brownkat6 committed:
isFinished() executes if there is no longer a tip problem

brownkat6 committed:
Moved AntiTip into Drive command and added RobotMath class

brownkat6 committed:
Added button to toggle whether AntiTip is enabled or not

brownkat6 committed:
Added SmartDashboard boolean to show if Anti-Tip is disabled

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -15,24 +20,60 @@
+                this.correctForPitch(stickSpeeds);
+            }
+        }
+        this.displayTippiness();
```
Repeated calls to `SmartDashboard.putThing` have been known to cause problems this year. Possibly due to limited bandwidth, but at any rate, be cautious of these. (The driver can see the robot tip most of the time... )

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -15,24 +20,60 @@
         requires(Drivetrain.getInstance());
     }
 
+    private void correctForPitch(double[] stickSpeeds) {//x-tip
```
Would it be better form to return the new list instead of modifying the one passed? I may be stuck in C++ thinking... :thinking:

c-x-berger replied:
It works. :man_shrugging:

c-x-berger commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -158,6 +160,14 @@ public double getUltrasonicDistance() {
+        return this.isAntiTipDisabled;
+    }
+
+    public void toggleAntiTip(){
```
Formatting nitpick... CTRL+ALT+L to avoid painful `git` troubles of the future.

c-x-berger commented on src/main/java/frc/robot/util/RobotMath.java:
```diff
@@ -0,0 +1,27 @@
+package frc.robot.util;
+
+public class RobotMath {
+    public static double powerCurve(double x) {
```
This does not belong here. As is, it causes confusion with the `powerCurve` method in `TwistArm`. They do similar things, but have entirely different implementations for doing so, and putting one of them in this `public` place causes enough confusion to be setting off red flags for me. 🚩

The appropriate place for this method is a `private static` member of `Drive`.

c-x-berger commented on src/main/java/frc/robot/util/RobotMath.java:
```diff
@@ -0,0 +1,27 @@
+        return Math.copySign(curve, x);
+    }
+
+    public static void normalize(double[] array) {
```
This function needs documentation badly. As is, I have no idea what it does (OK, after reading it a few times I get a picture, but still.)

Also, an array parameter named `array` is... less than ideal. This falls under the same umbrella as documenting the method to begin with.

c-x-berger replied:
Also, is this ever useful outside of `Drive`? (another thing documentation would help.) If not... see comments on moving of `powerCurve`.

c-x-berger commented on src/main/java/frc/robot/util/RobotMath.java:
```diff
@@ -0,0 +1,27 @@
+        double max = Math.abs(array[0]);
+        boolean normFlag = max > 1;
+
+        for (int i = 1; i < array.length; i++) {
```
Why not iterate over the whole array if we basically treat `array[0]` the same as the rest of the items? You could initialize `max` as 0, `normFlag` as `false`, then make this loop much cleaner as `for (double reasonableName : array) {`.

c-x-berger commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -172,6 +182,7 @@ public void periodic() {
                     + PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.LEFT_FOLLOWER_ONE_CHANNEL) + ", " +
                     PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.LEFT_FOLLOWER_TWO_CHANNEL));
         }
+        SmartDashboard.putBoolean("Anti-Tip Enabled", !this.isAntiTipDisabled);
```
Previous comment about repeated `SmartDashboard`ing applies here too.

brownkat6 committed:
Moved RobotMath functions to Drive

Step 1 of appeasing Caleb's wrath

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -15,24 +20,60 @@
+    }
+
+    private void updatePitchStatus() {
+        this.pitchDegrees = NavX.getInstance().getPitchDegrees();
```
It seems saner to just do this assignment in `execute` directly (less jumping around both in reading and, as a lesser problem, in actual execution)

c-x-berger reviewed (changes requested):
- Documentation needed on `normalize`
- Return `powerCurve` to its rightful place
- See if you can do something about spamming the SmartDashboard?

Good call on the button to toggle anti-tip.

brownkat6 committed:
added normalize() documentation

Step 2 of appeasing Caleb's wrath

brownkat6 committed:
smartdashboard antiTipEnabled boolean only updated when its toggled

brownkat6 committed:
commented out SmartDashboard boolean showing pitch

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -26,6 +23,34 @@ public Drive() {
+     * If any of the values are more than 1, they aren't valid values for motor power.
+     * If so, it divides all array values by the largest value to preserve the value ratios while making them valid motor power values.
+    */
+    public static void normalize(double[] array) {
```
Better placement, and thank you for the documentation. I still feel less-than-happy with the parameter name, but I wouldn't block merge for that alone (other, stricter people certainly would though, because it's painfully non descriptive.)

c-x-berger replied:
Also, I'll fix this myself later, but please read up on how to write [Javadoc.](https://en.wikipedia.org/wiki/Javadoc#Overview_of_Javadoc)

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -26,6 +23,34 @@ public Drive() {
         requires(Drivetrain.getInstance());
     }
 
+    public static double powerCurve(double x) {
```
Still `public`, which is kinda sketch. But at least it's a member of the right class now!

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -15,24 +17,88 @@
+        //if its over 45 there's no point in correcting it
+        if (Math.abs(pitchDegrees) < 45) {
+            //correctionFactor keeps the tilt correction within a certain threshold so it doesn't correct too much
+            double correctionFactor = (RobotMap.DRIVE.MAX_CORRECTION_FACTOR - RobotMap.DRIVE.MIN_CORRECTION_FACTOR) / (45 - RobotMap.DRIVE.PITCH_THRESHOLD_DEGREES);
```
This smells of suspicious magic. I'll look into how it works later. (Not merge blocking, but copying formulas without understanding them is not really the best idea. Of course, I would _never_ do such a thing... /s)

Also, is this not constant? If it is, why not make it a `final` field of something (such as this very class or `RobotMap.DRIVE`)?

c-x-berger reviewed (commented):
Much improved, thank you. I think I still see some places where you need to format your code (TODO: make that a failable Travis check), but it otherwise is mostly OK.

brownkat6 committed:
made normalize() and powerCurve() public

brownkat6 committed:
moved correction_factor constant to RobotMap insttead from Drive

brownkat6 committed:
added CORRECTION_FACTOR constant

brownkat6 committed:
Merge remote-tracking branch 'origin/anti-tip' into anti-tip

c-x-berger commented on src/main/java/frc/robot/commands/drive/Drive.java:
```diff
@@ -23,30 +23,31 @@ public Drive() {
     }
 
-    public static double powerCurve(double x) {
+    private double powerCurve(double x) {
```
Should be `static`

c-x-berger reviewed (commented):
What's left is nitpicky enough that I don't feel bothered fixing it myself.

- `powerCurve` doesn't need access to `this` and should thus be `static`.
  - Thank you for addressing the access of that though
- Good job addressing the parameter naming problem with `normalize`

c-x-berger committed:
Merge branch 'develop' into anti-tip

c-x-berger committed:
Make powerCurve static

c-x-berger committed:
General code formatting because I'm a nitpicker

c-x-berger committed:
Remove unused/commented out code

We have version control for this y'know

c-x-berger committed:
Slight tweak to tip flow control 

- Don't update pitch unless we need to
- Fetch pitch (hopefully) exactly one times per execute call
Looking at this more, maybe pitch should have been a parameter of correctForPitch?

c-x-berger commented:
This is now pending hardware testing for merge.

Commit:
Anti-tip updates

Commit:
removed pitchDegree loop

The if statement that prevented pitchDegrees from ever going negative was unnecessary. It also made it much more confusing to know whether the robot should correct for the tip, because instead of doing Math.abs(pitchDegrees), you had to do (pitchDegrees < threshold || pitch Degrees > (360-threshold).

Commit:
anti-tip edits speed when pitch is low, and takes when its high

Commit:
Removed debugging SmartDashboard booleans

c-x-berger reviewed (approved):
This all looks good, even after kludging it a bit for fitting the practice bot. Good job :+1:

I'll fix some of the formatting nitpicks myself and merge this ASAP.

c-x-berger committed:
Nittiest of nitpicks
